From b8917e552eaef8947459f4b8a6807f0b19697c9d Mon Sep 17 00:00:00 2001
From: Ethan Chen <ethan84@andestech.com>
Date: Thu, 24 Aug 2023 14:42:49 +0800
Subject: [PATCH 070/241] atcrtc100: Support atcrtc100 on AE350

Some functions are not supported in QEMU:
* Alarm Wakeup signal: Not used on FPGA
* Hardware Digital Trimming & Frequency Test: assume RTC clock  is
   accurate in emulation

Signed-off-by: Ethan Chen <ethan84@andestech.com>
---
 hw/riscv/Kconfig               |   1 +
 hw/riscv/andes_ae350.c         |   8 +-
 hw/rtc/Kconfig                 |   3 +
 hw/rtc/atcrtc100.c             | 408 +++++++++++++++++++++++++++++++++
 hw/rtc/meson.build             |   1 +
 include/hw/riscv/andes_ae350.h |   2 +
 include/hw/rtc/atcrtc100.h     |  30 +++
 7 files changed, 451 insertions(+), 2 deletions(-)
 create mode 100644 hw/rtc/atcrtc100.c
 create mode 100644 include/hw/rtc/atcrtc100.h

diff --git a/hw/riscv/Kconfig b/hw/riscv/Kconfig
index 1ccb60c694..93374588f6 100644
--- a/hw/riscv/Kconfig
+++ b/hw/riscv/Kconfig
@@ -14,6 +14,7 @@ config ANDES_AE350
     select ANDES_PLMT
     select ATCDMAC300
     select ATCPIT100
+    select ATCRTC100
     select ATCSMU
     select ATFMAC100
     select ATFSDC010
diff --git a/hw/riscv/andes_ae350.c b/hw/riscv/andes_ae350.c
index afd13da699..acd6768865 100644
--- a/hw/riscv/andes_ae350.c
+++ b/hw/riscv/andes_ae350.c
@@ -49,6 +49,7 @@
 #include "hw/riscv/andes_ae350.h"
 #include "hw/riscv/andes_atcsmu.h"
 #include "hw/sd/atfsdc010.h"
+#include "hw/rtc/atcrtc100.h"
 
 #define BIOS_FILENAME ""
 
@@ -408,8 +409,11 @@ static void andes_ae350_soc_realize(DeviceState *dev_soc, Error **errp)
         memmap[ANDES_AE350_WDT].base, memmap[ANDES_AE350_WDT].size);
 
     /* RTC */
-    create_unimplemented_device("riscv.andes.ae350.rtc",
-        memmap[ANDES_AE350_RTC].base, memmap[ANDES_AE350_RTC].size);
+    atcrtc100_create(memmap[ANDES_AE350_RTC].base,
+                     qdev_get_gpio_in(DEVICE(s->plic),
+                     ANDES_AE350_RTC_PERIOD_IRQ),
+                     qdev_get_gpio_in(DEVICE(s->plic),
+                     ANDES_AE350_RTC_ALARM_IRQ));
 
     /* GPIO */
     create_unimplemented_device("riscv.andes.ae350.gpio",
diff --git a/hw/rtc/Kconfig b/hw/rtc/Kconfig
index d0d8dda084..cf95fc72ce 100644
--- a/hw/rtc/Kconfig
+++ b/hw/rtc/Kconfig
@@ -1,3 +1,6 @@
+config ATCRTC100
+    bool
+
 config DS1338
     bool
     depends on I2C
diff --git a/hw/rtc/atcrtc100.c b/hw/rtc/atcrtc100.c
new file mode 100644
index 0000000000..e4f95545f5
--- /dev/null
+++ b/hw/rtc/atcrtc100.c
@@ -0,0 +1,408 @@
+/*
+ * Andes Real-time Clcok, ATCRTC100.
+ *
+ * Copyright (c) 2023 Andes Tech. Corp.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#include "qemu/osdep.h"
+#include "qemu/bitops.h"
+#include "qemu/log.h"
+#include "qapi/error.h"
+#include "qemu/error-report.h"
+#include "hw/irq.h"
+#include "hw/ptimer.h"
+#include "hw/qdev-properties.h"
+#include "hw/sysbus.h"
+#include "hw/rtc/atcrtc100.h"
+
+#define xLOG(x...)
+#define yLOG(x...) qemu_log(x)
+#define zLOG(x...) printf(x)
+
+#define LOG(x...) xLOG(x)
+#define LOGGE(x...) qemu_log_mask(LOG_GUEST_ERROR, x)
+
+#define DEVNAME "ATCRTC100"
+
+enum {
+    ID_RTC = 0x030110,
+    ID_RTC_SHIFT = 8,
+};
+
+enum {
+    VER_MAJOR = 0x00,
+    VER_MAJOR_SHIFT = 4,
+    VER_MINOR = 0x00,
+    VER_MINOR_SHIFT = 0,
+};
+
+enum {
+    REG_IDREV = 0x00,
+    REG_CNTR = 0x10,
+    REG_ALARM = 0x14,
+    REG_CTRL = 0x18,
+    REG_ST = 0x1c,
+    REG_TRIM = 0x20,
+};
+
+enum {
+    CNTR_DAY = 17,
+    CNTR_HOUR = 12,
+    CNTR_MIN = 6,
+    CNTR_SEC = 0,
+};
+
+enum {
+    CTRL_FREQ_TEST_EN = 8,
+    CTRL_HSEC = 7,
+    CTRL_SEC = 6,
+    CTRL_MIN = 5,
+    CTRL_HOUR = 4,
+    CTRL_DAY = 3,
+    CTRL_ALARM_INT = 2,
+    CTRL_ALARM_WACKUP = 1,
+    CTRL_RTC_EN = 0,
+};
+
+enum {
+    ST_WRITEDONE = 16,
+    ST_HSEC = 7,
+    ST_SEC = 6,
+    ST_MIN = 5,
+    ST_HOUR = 4,
+    ST_DAY = 3,
+    ST_ALARM_INT = 2,
+};
+
+static inline uint64_t
+tm2ns(struct tm tm)
+{
+    return (((((((uint64_t)tm.tm_mday * 24)
+              + tm.tm_hour) * 60)
+              + tm.tm_min) * 60)
+              + tm.tm_sec) * NANOSECONDS_PER_SECOND;
+}
+
+static inline struct tm
+ns2tm(uint64_t ns)
+{
+    struct tm tm;
+    tm.tm_sec = ns / NANOSECONDS_PER_SECOND;
+    tm.tm_min = tm.tm_sec / 60;
+    tm.tm_sec %= 60;
+    tm.tm_hour = tm.tm_min / 60;
+    tm.tm_min %= 60;
+    tm.tm_mday = tm.tm_hour / 24;
+    return tm;
+}
+
+static inline struct tm
+cntr2tm(uint32_t cntr)
+{
+    struct tm tm;
+    tm.tm_mday = (cntr >> CNTR_DAY);
+    tm.tm_hour = (cntr >> CNTR_HOUR) & 0x3f;
+    tm.tm_min = (cntr >> CNTR_MIN) & 0x3f;
+    tm.tm_sec = cntr & 0x3f;
+    return tm;
+}
+
+static inline uint32_t
+tm2cntr(struct tm tm)
+{
+    return tm.tm_mday << CNTR_DAY | tm.tm_hour << CNTR_HOUR |
+           tm.tm_min << CNTR_MIN | tm.tm_sec;
+}
+
+static inline uint32_t
+ns2cntr(uint64_t ns)
+{
+    return tm2cntr(ns2tm(ns));
+}
+
+static void
+atcrtc100_reset(DeviceState *dev)
+{
+    Atcrtc100State *s = ATCRTC100(dev);
+    s->cntr = 0;
+    s->alarm = 0;
+    s->ctrl = 0;
+    s->st = (1 << ST_WRITEDONE);
+    s->trim = 0;
+    s->ref_rtc_start_ns = 0;
+    s->pause_ns = 0;
+    s->period_tm = cntr2tm(0);
+    LOG("%s:\n", __func__);
+}
+
+static void
+atcrtc100_realize(DeviceState *dev, Error **errp)
+{
+    LOG("%s:\n", __func__);
+}
+
+static uint64_t
+read_atcrtc100_cntr(Atcrtc100State *s)
+{
+    if (s->ctrl & (1 << CTRL_RTC_EN)) {
+        return ns2cntr(
+                qemu_clock_get_ns(QEMU_CLOCK_REALTIME)
+                - s->ref_rtc_start_ns);
+    } else {
+        return ns2cntr(s->pause_ns);
+    }
+}
+
+static void
+write_atcrtc100_cntr(Atcrtc100State *s, uint64_t value)
+{
+    struct tm value_tm = cntr2tm(value);
+    s->period_tm = value_tm;
+    s->ref_rtc_start_ns = qemu_clock_get_ns(QEMU_CLOCK_REALTIME)
+                          - tm2ns(value_tm);
+    s->pause_ns = 0;
+}
+
+static inline uint64_t
+next_half_second_ns(uint64_t ns)
+{
+    return (((ns + (NANOSECONDS_PER_SECOND / 2)) / NANOSECONDS_PER_SECOND))
+           * NANOSECONDS_PER_SECOND;
+}
+
+static void
+write_atcrtc100_ctrl(Atcrtc100State *s, uint64_t value)
+{
+    if (value & (1 << CTRL_RTC_EN) && !(s->ctrl & (1 << CTRL_RTC_EN))) {
+        /* resume */
+        s->ref_rtc_start_ns = qemu_clock_get_ns(QEMU_CLOCK_REALTIME)
+                              - s->pause_ns;
+    } else if ((!(value & (1 << CTRL_RTC_EN))) &&
+               (s->ctrl & (1 << CTRL_RTC_EN))) {
+        /* pause */
+        s->pause_ns = qemu_clock_get_ns(QEMU_CLOCK_REALTIME)
+                      - s->ref_rtc_start_ns;
+    }
+    if (value & (1 << CTRL_RTC_EN) && value & (0xFE)) {
+        uint64_t ns = qemu_clock_get_ns(QEMU_CLOCK_REALTIME)
+                      - s->ref_rtc_start_ns;
+        timer_mod(s->rtc_timer, next_half_second_ns(ns) + s->ref_rtc_start_ns);
+    }
+    s->ctrl = value;
+}
+
+static uint64_t
+atcrtc100_read(void *opaque, hwaddr addr, unsigned size)
+{
+    Atcrtc100State *s = ATCRTC100(opaque);
+    uint64_t rz = 0;
+
+    switch (addr) {
+    case REG_IDREV ... REG_TRIM:
+        switch (addr) {
+        case REG_IDREV: /* RO */
+            rz = ID_RTC << ID_RTC_SHIFT | VER_MAJOR << VER_MAJOR_SHIFT |
+                 VER_MINOR << VER_MINOR_SHIFT;
+            break;
+        case REG_CNTR:
+            rz = read_atcrtc100_cntr(s);
+            break;
+        case REG_ALARM:
+            rz = s->alarm;
+            break;
+        case REG_CTRL:
+            rz = s->ctrl;
+            break;
+        case REG_ST:
+            rz = s->st;
+            break;
+        case REG_TRIM:
+            rz = s->trim;
+            break;
+
+        default:
+            LOGGE("%s: Bad addr %x\n", __func__, (int)addr);
+            break;
+        }
+        LOG("\e[95m%s: addr %08x, value %08x\e[0m\n", __func__, (int)addr,
+            (int)rz);
+        break;
+    default:
+        LOGGE("%s: Bad addr %x\n", __func__, (int)addr);
+    }
+
+    return rz;
+}
+
+
+static void
+atcrtc100_write(void *opaque, hwaddr addr, uint64_t value, unsigned size)
+{
+    Atcrtc100State *s = ATCRTC100(opaque);
+
+    s->st &= ~(1 << ST_WRITEDONE);
+    switch (addr) {
+    case REG_IDREV ... REG_TRIM:
+        switch (addr) {
+        case REG_IDREV: /* RO */
+            break;
+        case REG_CNTR:
+            write_atcrtc100_cntr(s, value);
+            break;
+        case REG_ALARM:
+            s->alarm = value;
+            break;
+        case REG_CTRL:
+            write_atcrtc100_ctrl(s, value);
+            break;
+        case REG_ST:
+            s->st &= ~(value & 0xFC);
+            break;
+        case REG_TRIM:
+            s->trim = value;
+            break;
+
+        default:
+            LOGGE("%s: Bad addr %x\n", __func__, (int)addr);
+            break;
+        }
+        LOG("\e[95m%s: addr %08x, value %08x\e[0m\n", __func__, (int)addr,
+            (int)rz);
+        break;
+    default:
+        LOGGE("%s: Bad addr %x\n", __func__, (int)addr);
+    }
+    s->st |= (1 << ST_WRITEDONE);
+}
+
+static void
+atcrtc100_timer_cb(void *opaque)
+{
+    Atcrtc100State *s = ATCRTC100(opaque);
+    uint64_t curr_ns = qemu_clock_get_ns(QEMU_CLOCK_REALTIME)
+                       - s->ref_rtc_start_ns;
+    struct tm curr_tm = ns2tm(curr_ns);
+    timer_mod(s->rtc_timer, next_half_second_ns(curr_ns) + s->ref_rtc_start_ns);
+    uint period_level = 0;
+    uint alarm_level = 0;
+    if (s->ctrl & (1 << CTRL_HSEC)) {
+        period_level |= 1;
+        s->st |= (1 << ST_HSEC);
+    }
+    if (s->ctrl & (1 << CTRL_SEC)) {
+        if (curr_tm.tm_sec != s->period_tm.tm_sec) {
+            period_level |= 1;
+            s->st |= (1 << ST_SEC);
+        }
+    }
+    if (s->ctrl & (1 << CTRL_MIN)) {
+        if (curr_tm.tm_min != s->period_tm.tm_min) {
+            period_level |= 1;
+            s->st |= (1 << ST_MIN);
+        }
+    }
+    if (s->ctrl & (1 << CTRL_HOUR)) {
+        if (curr_tm.tm_hour != s->period_tm.tm_hour) {
+            period_level |= 1;
+            s->st |= (1 << ST_HOUR);
+        }
+    }
+    if (s->ctrl & (1 << CTRL_DAY)) {
+        if (curr_tm.tm_mday != s->period_tm.tm_mday) {
+            period_level |= 1;
+            s->st |= (1 << ST_DAY);
+        }
+    }
+    s->period_tm = curr_tm;
+
+    if (s->ctrl & (1 << CTRL_ALARM_INT)) {
+        /* Only trigger alarm in first half second */
+        if (tm2cntr(curr_tm) == s->alarm && (curr_ns % NANOSECONDS_PER_SECOND)
+            < NANOSECONDS_PER_SECOND / 2) {
+            alarm_level = 1;
+            s->st |= (1 << ST_ALARM_INT);
+        }
+    }
+    if (period_level) {
+        qemu_set_irq(s->period_irq, period_level);
+    }
+    if (alarm_level) {
+        qemu_set_irq(s->alarm_irq, alarm_level);
+    }
+}
+
+static const MemoryRegionOps atcrtc100_ops = {
+    .read = atcrtc100_read,
+    .write = atcrtc100_write,
+    .endianness = DEVICE_NATIVE_ENDIAN,
+    .valid = {.min_access_size = 4, .max_access_size = 4}
+};
+
+
+static void
+atcrtc100_class_init(ObjectClass *klass, void *data)
+{
+    LOG("%s:\n", __func__);
+    DeviceClass *k = DEVICE_CLASS(klass);
+    k->realize = atcrtc100_realize;
+    k->reset = atcrtc100_reset;
+}
+
+static void
+atcrtc100_init(Object *obj)
+{
+    LOG("%s:\n", __func__);
+    Atcrtc100State *s = ATCRTC100(obj);
+    SysBusDevice *sbd = SYS_BUS_DEVICE(obj);
+    sysbus_init_irq(sbd, &s->period_irq);
+    sysbus_init_irq(sbd, &s->alarm_irq);
+    s->st = (1 << ST_WRITEDONE);
+    s->rtc_timer = timer_new_ns(QEMU_CLOCK_REALTIME, atcrtc100_timer_cb, obj);
+    memory_region_init_io(&s->mmio, obj, &atcrtc100_ops, s, TYPE_ATCRTC100,
+                          0x100);
+    sysbus_init_mmio(sbd, &s->mmio);
+}
+
+static const TypeInfo atcrtc100_info = {
+    .name = TYPE_ATCRTC100,
+    .parent = TYPE_SYS_BUS_DEVICE,
+    .instance_size = sizeof(Atcrtc100State),
+    .instance_init = atcrtc100_init,
+    .class_init = atcrtc100_class_init,
+};
+
+static void
+atcrtc100_register_types(void)
+{
+    LOG("%s:\n", __func__);
+    type_register_static(&atcrtc100_info);
+}
+
+type_init(atcrtc100_register_types)
+
+DeviceState *
+atcrtc100_create(hwaddr addr, qemu_irq period_irq, qemu_irq alarm_irq)
+{
+    LOG("%s:\n", __func__);
+    DeviceState *dev;
+    dev = sysbus_create_varargs(TYPE_ATCRTC100, addr, period_irq, alarm_irq,
+                                NULL);
+    return dev;
+}
diff --git a/hw/rtc/meson.build b/hw/rtc/meson.build
index 3ea2affe0b..e32e15ce39 100644
--- a/hw/rtc/meson.build
+++ b/hw/rtc/meson.build
@@ -14,3 +14,4 @@ system_ss.add(when: 'CONFIG_GOLDFISH_RTC', if_true: files('goldfish_rtc.c'))
 system_ss.add(when: 'CONFIG_LS7A_RTC', if_true: files('ls7a_rtc.c'))
 system_ss.add(when: 'CONFIG_ALLWINNER_H3', if_true: files('allwinner-rtc.c'))
 system_ss.add(when: 'CONFIG_MC146818RTC', if_true: files('mc146818rtc.c'))
+system_ss.add(when: 'CONFIG_ATCRTC100', if_true: files('atcrtc100.c'))
diff --git a/include/hw/riscv/andes_ae350.h b/include/hw/riscv/andes_ae350.h
index 057fc67528..8bad146d87 100644
--- a/include/hw/riscv/andes_ae350.h
+++ b/include/hw/riscv/andes_ae350.h
@@ -111,6 +111,8 @@ enum {
 };
 
 enum {
+    ANDES_AE350_RTC_PERIOD_IRQ = 1,
+    ANDES_AE350_RTC_ALARM_IRQ = 2,
     ANDES_AE350_PIT_IRQ = 3,
     ANDES_AE350_UART1_IRQ = 8,
     ANDES_AE350_UART2_IRQ = 9,
diff --git a/include/hw/rtc/atcrtc100.h b/include/hw/rtc/atcrtc100.h
new file mode 100644
index 0000000000..63d95ecb35
--- /dev/null
+++ b/include/hw/rtc/atcrtc100.h
@@ -0,0 +1,30 @@
+#ifndef HW_ATCRTC100_H
+#define HW_ATCRTC100_H
+
+#include "hw/sysbus.h"
+#include "qemu/typedefs.h"
+#define TYPE_ATCRTC100 "atcrtc100"
+#define ATCRTC100(obj) OBJECT_CHECK(Atcrtc100State, (obj), TYPE_ATCRTC100)
+
+typedef struct Atcrtc100State {
+    SysBusDevice parent_obj;
+
+    MemoryRegion mmio;
+    qemu_irq period_irq;
+    qemu_irq alarm_irq;
+    int64_t ref_rtc_start_ns;
+    int64_t pause_ns;
+    struct tm period_tm;
+    QEMUTimer *rtc_timer;
+
+    uint32_t id_rev;
+    uint32_t cntr;
+    uint32_t alarm;
+    uint32_t ctrl;
+    uint32_t st;
+    uint32_t trim;
+} Atcrtc100State;
+
+DeviceState *atcrtc100_create(hwaddr addr, qemu_irq period_irq,
+                              qemu_irq alarm_irq);
+#endif /* HW_ATCRTC100_H */
-- 
2.34.1

